### ISSUE TITLE
Model selection

### ISSUE THREAD

**Issue by manhminno** [microsoft/ProDA#24]
Why do you use validation dataset (500 images Cityscapes) to select best model? Valid dataset (500 images Cityscapes) is only used for evaluation, not for model selection (for saving best model), this paper is having a big problem!

**Comment by manhminno** [comment]
The results of this paper are very high, I think because you save the best model on 500 Cityscapes (model selection), and use that model to evaluate and save this results. That's not fair!

**Comment by panzhang0104** [comment]
The validation dataset (not test dataset) is used to choose the hyperparameters, including learning rate, batch size, training iterations.  It is fair to compare the performance on test dataset. However, only a few UDA papers report performance on val dataset and test dataset, most only report validation dataset. And these works use val dataset to select hyperparameters ( especial training iterations). For e.g [CAG_UDA](https://github.com/RogerZhangzz/CAG_UDA) and [Seg_Uncertainty](https://github.com/layumi/Seg-Uncertainty). The evaluation of this work is borrowed from CAG_UDA.

**Comment by panzhang0104** [comment]
In our experiment, even we do not use best model, and only use the last model, the performance only drop <1.0 mIoU, still much better than previous work. And the ablation study in the paper use same setting.

**Comment by panzhang0104** [comment]
If you have any other question, please email me.

**Comment by manhminno** [comment]
Dear @panzhang0212 ,
Sorry for my late reply. In your source code, you saved the best model based on the evaluation on the valid set (500 images). Then use that model in generate_pseudo_label stage to generate pseudo label and then evaluation on the valid set to saving best model and then that model continues to be used in denoising stage. During the training process, you always use the evaluation on the valid set (500 images) to select the model.
Only use the valid set to see the results (make it being unseen dataset), not base on it to select the model and tuning model.

**Comment by panzhang0212** [comment]
In our experiment, using the best model or current model to generate pseudo labels performs almost the same, even the quality of the pseudo label is a little bit different. But the gain(<1mIoU) of the pseudo label generated by the best model is not enough to help the next stage training. We have done the above experiments. The models trained by two sets of pseudo labels (performance < 2 mIoU) usually performance similar. And you can try the code without using the best model in any stages. The performance is similiar.

**Comment by manhminno** [comment]
Thanks for your immediate reply. Can you publish the results without using the valid dataset to select and save the best model?
You use the valid dataset (500 images) to select model and save the best model. It has influenced from the first process (warm-up), then continues to do this for the process of denoising stage, knowledge distillation stage. The whole training process depends on the valid dataset (500 images). If you use this valid set as unseen dataset (only to see the final result), the final result will probably drop a lot, 57.5% mIoU of result is impossible. Even the result will be bad, we need to verify this.

**Comment by panzhang0212** [comment]
Why do you think the result is impossible? And the definition of the validation dataset in the wiki is 
![image](https://user-images.githubusercontent.com/30363822/123884982-0f8e7380-d97f-11eb-8efa-6a05385d4c40.png)
We do not use the **test dataset** to do model selection, so I think there is no problem (In fact, most of the UDA work use validation set to select the best iteration as the final iteration, which is similar to model selection).  

There are two ways to dispel your worry. First, Using our release model to test on the test dataset (the test dataset is not public). Second, changing the name of the validation set to test set, so anyone could not use it for hyperparameters and model selection (In the way, we need to do experiments using the current model, not the best model, and when I am free, I will conduct these experiments ). But the second way is not suitable, validation set is validation set, test dataset is test dataset

**Comment by panzhang0212** [comment]
If you are interested in the setting, you can learn more detail from other UDA work, e.g [CAG_UDA](https://github.com/RogerZhangzz/CAG_UDA). The reason we using the best model is that we follow the UDA work from the released code of CAG_UDA which uses the best warm-up model. I remember changing the best model to the current model only drop a little or no influence. If you are interested current/best model, I will conduct experiments using current model when I'm free.

**Comment by saolaisa02** [comment]
I am also thinking about this issue and I have the same opinion like @manhminno 
Evaluating the model on validation dataset (500 images cityscapes) and then choosing to save the best model on that set for use in subsequent processes has no difference from giving validation dataset as training data. In that case, evaluating on validation dataset and then getting the results for this paper no longer makes sense. The best way is to consider set validation dataset as test dataset - not validation dataset - only to see result. After the process of training the model is totally completed, we are going to evaluate it (500 images - validation dataset) and give the results.

**Comment by super233** [comment]
I'm also curious about this issue. Although the 57.5% MIoU of result is unbelievable, I think the gain in stage 2\~3 is mainly caused by the self-training with segmentation loss in the source and target domain, but there's a lack of ablation study for that only self-training and only knowledge distillation are available separately in stage 2\~3, can you provide the ablation study if you have done before @panzhang0212 ?

By the way, can this issue be reopened @panzhang0212 and give a readable title @manhminno so that others can join the discussion?

**Comment by panzhang0212** [comment]
The ablations are shown in Table 3 of the paper, the gain is mainly caused by prototypical denoising, what other ablations do you want to know? @super233
![image](https://user-images.githubusercontent.com/30363822/123936558-ddedca80-d9c7-11eb-8959-9db4ec02b46f.png)

**Comment by super233** [comment]
Thanks for your immediate reply. 😄 I want to know the ablation study of that only the self-training and only the knowledge distillation are avaliable in stage 2~3 separately. There is only the ablation study of different init method for knowledge distillation in table 3.

**Comment by zhangmozhe** [comment]
Thanks for the question and let us clarify this.

First, our method (red curve) steadily improves the performance as shown in the below figure. Compared to the baseline (self-training shown as blue curve), we can clearly see the substantial performance boost. The performance is stable in the late training phase of 3rd stage. That is, the model selection does not bring much influence in the final score.
![image](https://user-images.githubusercontent.com/10203551/123933000-8c900c00-d9c4-11eb-8bc2-10dabd3890d1.png)

Second, our evaluation approach follows the convention in this area. In fact, the work [CAG_UDA](https://github.com/RogerZhangzz/CAG_UDA) and [Seg_Uncertainty](https://github.com/layumi/Seg-Uncertainty) use the same way for the evaluation. [AdaSeg](https://github.com/wasidennis/AdaptSegNet) and [BDL](https://github.com/liyunsheng13/BDL) implicitly choose the model with early stop and the number of training steps is manually specified. So the comparison in our paper is fair in terms of experiment settings, which shows the state-of-the-art performance of our method.

Third, we do not agree that this setting is equivalent to leverage evaluation data as training data. This is because domain adaptation is a special scenario, to report the mIoU in this way does make sense. Only in this way, we can conduct comparisons against prior works. 

Finally, we agree that better evaluation metrics can be proposed. For example, temporal average mIoU score of multiple checkpoints as well as the performance variance. We welcome future works use such measures and unify the evaluation protocol in this area. Still, as we visualize, our performance is stable towards the end of the training phase, ~57%. In contrast, adversarial domain adaptation, in our experience, suffers much stronger mIoU instability during training. The model selection is a more severe problem for adversarial-based approach rather than self-training approaches.

We hope our answer clarifies the concern and we believe the experiments and conclusions in the paper are fair and solid. In terms of more ablation studies, welcome to train our code from scratch. The performance is reproducible if following the same training script. If you have further questions about the ablation, please open a new issue.

**Comment by panzhang0212** [comment]
The stage1 is denoising stage, and  stage2~3 (no denoising) are self-training which utilizes self-supervised pretrain model. However, if we apply stage1 with self-training, the strategy is similar to the general self-training paper (w/o self-distillation). And the self-distillation could only improve about 1 mIoU, so the key of this paper is denoising.

In fact, we could apply denosing strategy to stage2 ~ 3, the performance is similar.  But self-training in stage2~3 could be trained faster than denosing stage. 

If you are interested in more ablation, you could open a new issue to discuss more ablation.@super233

**Comment by saolaisa02** [comment]
Can you re-open this issue? I think a lot of people are wondering about this too.

**Comment by super233** [comment]
Hi, which result is reported in ablation study for each training setting, the last iteration or the best iteration? :-)
